crates/block-producer/src/block_builder/mod.rs

```diff
@@ -72,37 +74,153 @@ impl BlockBuilder {
         loop {
             interval.tick().await;
 
-            let (block_number, batches) = mempool.lock().await.select_block();
+            self.build_block(&mempool).await;
+        }
+    }
 
-            let mut result = self.build_block(&batches).await;
-            let proving_duration = rand::thread_rng().gen_range(self.simulated_proof_time.clone());
+    #[instrument(parent = None, target = COMPONENT, name = "block_builder.build_block", skip_all)]
+    async fn build_block(&self, mempool: &SharedMempool) {
+        use futures::TryFutureExt;
+
+        // Run the block building stages and add open-telemetry trace information where applicable.
+        //
+        // Important telemetry is added to the root span, and stages are free to add additional data
+        // to their subspans. Each stage has its own span and is expected to emit an error
+        // event if it fails. This automatically marks the span as an error, however the
+        // root span must be manually marked as this function does not return a result.
+        Self::select_block(mempool)
+            .inspect(|selected| selected.inject_telemetry(Span::current()))
+            .then(|selected| self.get_block_inputs(selected))
+            .inspect_ok(|inputs| inputs.inject_telemetry(Span::current()))
+            .and_then(|inputs| self.prove_block(inputs))
+            .inspect_ok(|proven_block| proven_block.inject_telemetry(Span::current()))
```

Comment:
If this isn't legible enough then we could refactor to something like:

```suggestion
            .inspect(SelectedBlock::inject_telemetry)
            .then(SelectedBlock::get_block_inputs)
            .inspect_ok(BlockPreimage::inject_telemetry)
            .and_then(BlockPreimage::prove_block)
            .inspect_ok(ProvenBlock::inject_telemetry)
```

by moving the `Span::curren()` into the `inject_telemetry` calls, and have the structs borrow `self` and `mempool`.

Though probably not worth it.